### ISSUE TITLE
[Docs] Cannot replicate documented RTMO throughput

### ISSUE THREAD

**Issue by btalberg** [open-mmlab/mmpose#2987]
### 📚 The doc issue

In your [RTMO introduction](https://github.com/open-mmlab/mmpose/tree/main/projects/rtmo), you state you achieved 141 FPS against a single V100 GPU with your RTMO-l model. I'm working on an inference-only engine that deploys your RTMO-l model. We've been attempting to achieve the same throughput on a V100, but have so far only achieved half that throughput (70-80 FPS). And that was only after a rigorous amount of tuning (e.g. using torch.compile, using FP16, using batching, moving image resizing and non-max suppression to torchvision, adding numba where possible).

Here's the repo I'm working on: https://github.com/WildflowerSchools/wf-pose-engine. Much of the pose inference/RTMO specific implementation details be found here: [/pose_engine/inference/pose_estimator.py](https://github.com/WildflowerSchools/wf-pose-engine/blob/main/pose_engine/inference/pose_estimator.py).

### Suggest a potential alternative/fix

I believe that you achieved 140 FPS throughput. But I'm wondering if you used any tweaks/customizations to hit that number. Could you shed any light on your configuration? Were you using your mm* libraries out of the box? Were you using batching, and if so, what batch size did you settle on? Did you have trouble getting images to and from the GPU fast enough? Any other "simple tricks" that helped you get more performance out of the V100?

Thanks for this project and any insight you can offer!

**Comment by Ben-Louis** [comment]
We utilize MMDeploy to export RTMO models and test inference speed via [this script](https://github.com/open-mmlab/mmdeploy/blob/main/tools/profiler.py). You may want to consider using MMDeploy directly.

**Comment by btalberg** [comment]
Thanks, @Ben-Louis. Beyond the GPU, can you share details about the box you did your speed testing against? Did you test against an AWS p3 instance? I'm running against a p3.2xlarge.

When I profile our code on the p3 instance, I notice moving tensors to the GPU performs slower than I'd expect.

For development, we use a GPU desktop w/ 2 RTX 2080s and an AMD Ryzen Threadripper 2950X CPU. The RTX 2080 GPU is meaningfully slower than a V100, but on our box, we get better _overall_ throughput and I believe it's because we can get tensors on the CPU over to the GPU much faster.

**CPU Tensor to GPU Throughput:** *(these are 600x600x3 shaped image tensors received from the pre-processing pipeline)*
* AWS p3.2xlarge (V100 + Xeon Skylake): 200 tensors/second
* Personal desktop (single RTX 2080 + Threadripper 2950X): 1,200 tensors/second

**Comment by btalberg** [comment]
FYI, for additional context to the comment I added above, I ran the `mmdeploy:tools/profiler.py` script against COCO's test2017 image dataset. I tested against an AWS p3.2xlarge V100 instance. I used the RTMO-l model with a batch size of 8. I got 54.7 FPS (mean). Here's the command I used and the output:

```
python tools/profiler.py \
    configs/mmpose/pose-detection_rtmo_onnxruntime_dynamic.py \
    /app/.venv/lib/python3.11/site-packages/mmpose/.mim/configs/body_2d_keypoint/rtmo/body7/rtmo-l_16xb16-600e_body7-640x640.py \
    /app/coco_images/test2017 \
    --model https://download.openmmlab.com/mmpose/v1/projects/rtmo/rtmo-l_16xb16-600e_body7-640x640-b37118ce_20231211.pth \
    --shape 640x640 \
    --device cuda \
    --num-iter 100 \
    --warmup 10 \
    --batch-size 8
```

> 03/18 17:06:11 - mmengine - WARNING - Failed to search registry with scope "mmpose" in the "Codebases" registry tree. As a workaround, the current "Codebases" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmpose" is a correct scope, or whether the registry is initialized.
> 03/18 17:06:11 - mmengine - WARNING - Failed to search registry with scope "mmpose" in the "mmpose_tasks" registry tree. As a workaround, the current "mmpose_tasks" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmpose" is a correct scope, or whether the registry is initialized.
> /app/.venv/lib/python3.11/site-packages/mmpose/datasets/datasets/utils.py:102: UserWarning: The metainfo config file "configs/_base_/datasets/coco.py" does not exist. A matched config file "/app/.venv/lib/python3.11/site-packages/mmpose/.mim/configs/_base_/datasets/coco.py" will be used instead.
>   warnings.warn(
> Loads checkpoint by http backend from path: https://download.openmmlab.com/mmpose/v1/projects/rtmo/rtmo-l_16xb16-600e_body7-640x640-b37118ce_20231211.pth
> 03/18 17:06:12 - mmengine - INFO - module neck has been switched to deploy mode
> /app/.venv/lib/python3.11/site-packages/torch/functional.py:507: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at ../aten/src/ATen/native/TensorShape.cpp:3549.)
>   return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
> 03/18 17:06:12 - mmengine - INFO - module neck.fpn_blocks.0.bottlenecks.0 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.fpn_blocks.0.bottlenecks.1 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.fpn_blocks.0.bottlenecks.2 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.fpn_blocks.1.bottlenecks.0 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.fpn_blocks.1.bottlenecks.1 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.fpn_blocks.1.bottlenecks.2 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.pan_blocks.0.bottlenecks.0 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.pan_blocks.0.bottlenecks.1 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.pan_blocks.0.bottlenecks.2 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.pan_blocks.1.bottlenecks.0 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.pan_blocks.1.bottlenecks.1 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module neck.pan_blocks.1.bottlenecks.2 has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module head has been switched to deploy mode
> 03/18 17:06:12 - mmengine - INFO - module head.dcc has been switched to deploy mode
> 03/18 17:06:13 - mmengine - INFO - Found totally 40670 image files in /app/coco_images/test2017
> 03/18 17:06:26 - mmengine - INFO - [pytorch]-30 times per count: 17.89 ms, 55.89 FPS
> 03/18 17:06:33 - mmengine - INFO - [pytorch]-50 times per count: 18.09 ms, 55.29 FPS
> 03/18 17:06:40 - mmengine - INFO - [pytorch]-70 times per count: 18.48 ms, 54.12 FPS
> 03/18 17:06:46 - mmengine - INFO - [pytorch]-90 times per count: 18.22 ms, 54.89 FPS
> 03/18 17:06:53 - mmengine - INFO - [pytorch]-110 times per count: 18.28 ms, 54.70 FPS
> ----- Settings:
> +------------+---------+
> | batch size |    12   |
> |   shape    | 640x640 |
> | iterations |   100   |
> |   warmup   |    10   |
> +------------+---------+
> ----- Results:
> +--------+------------+--------+
> | Stats  | Latency/ms |  FPS   |
> +--------+------------+--------+
> |  Mean  |   18.282   | 54.698 |
> | Median |   17.605   | 56.801 |
> |  Min   |   13.645   | 73.288 |
> |  Max   |   27.845   | 35.913 |
> +--------+------------+--------+

**Comment by btalberg** [comment]
Indeed, I was able to get faster throughput using TensorRT. My major bottleneck was because of video -> frames computation that I do before passing images to the inference engine. If I prepare the video as still frames/image before hand and then test my throughput, I can reach 240 FPS! The p3 instances just do not give me enough CPU power to both prepare video for inference AND keep up with the inference engine's potential.

So, I moved my application to a p5.4xlarge which gives me 16 CPUs and an A10G GPU. There I can reach ~150 FPS with my video -> frames step in tack.
